[자동차 경주 게임] y 미션 제출합니다.

jiyoung-Heo opened the pull request:
지난주 미션때 피드백받은 내용에 유의하여 작성하였습니다.   
특히 매직넘버 설정과 클래스 분리 부분에 더욱 신경쓰며 작성하였습니다!!   

- 이번 미션은 리드미 체크목록 하나당 커밋 하나를 진행하다보니 너무 많은 기능이 하나의 커밋에 포함된다고 느꼈습니다. 더욱 더 세세하게 리드미를 작성하도록 하겠습니다.    
- 항상 돌아가기만 하면 되는 코드를 짜다가 잘 짜여진 코드를 짜려고 하니 아직 부족한부분이 많다는것을 다시한번 깨달았습니다.   

Q. Validation 클래스와 CheckCarName, CheckPlayTime 클래스를 분리하였는데 validation에서 생성자를 통해 두 클래스의 메소드를 실행시켜도 괜찮은 것인지 궁금합니다. 

감사합니다 🙂

jiyoung-Heo committed:
docs(README.md): 구현할 기능 목록 작성

jiyoung-Heo committed:
feat(InputCarNAmes): 자동차 이름 입력받기

jiyoung-Heo committed:
feat(inputPlayTimes): 시도할 횟수 입력받기

jiyoung-Heo committed:
feat(checkCarNameLastComma): 자동차이름 입력 시 가장 마지막에 쉼표가 올 경우 예외처리 기능 추가

jiyoung-Heo committed:
feat(RacingGame): 게임을 시작하여 전진기능 구현

jiyoung-Heo committed:
feat(RacingGame): 각 차수별 실행 결과 출력

jiyoung-Heo committed:
feat(RacingGame): 모든 시도가 끝난 후 우승자 출력

jiyoung-Heo committed:
style(playOnce,playGame): 메소드명 변경

jiyoung-Heo committed:
style: 컨벤션 어긋나는부분 수정

jiyoung-Heo committed:
Refactor: 상수 수정 및 추가

jiyoung-Heo committed:
Refactor(validate): validate 패키지 추가

jiyoung-Heo committed:
style(PRINT_MESSAGE_FINAL_WINNER): 띄워쓰기 추가

xlffm3 commented on src/main/java/racingcar/Application.java:
```diff
@@ -1,7 +1,17 @@
-    public static void main(String[] args) {
-        // TODO 구현 진행
-    }
+	public static void main(String[] args) {
+		InputUser inputUser = new InputUser();
+		String[] inputCarName = inputUser.inputCarNames().split(",");
+
```
일반적으로 1 tab = 4 space인데요. Indent가 다르게 설정되어있는 것 같습니다.
확인 부탁드립니다~

jiyoung-Heo replied:
![image](https://user-images.githubusercontent.com/93931316/145941306-bcecb3f7-b949-4a4c-861b-b04d00425af8.png)
현재 이렇게 설정되어 있습니다! 
제 화면 상으로는 4space 되어 보이는데 다르게 보이시나요...??

xlffm3 replied:
![image](https://user-images.githubusercontent.com/56240505/147225709-b888d6cd-f1a0-479c-8f57-7fc902e65eae.png)

pr 리뷰상에서는 다르게 보이네요.

xlffm3 commented on src/main/java/racingcar/validate/Validation.java:
```diff
@@ -0,0 +1,14 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class Validation {
```
유효성을 검증해주는 클래스네요.
그런데 조금 더 명시적으로 'Validator' 같은 클래스 명은 어떨까요?
클래스, 메서드, 변수 명은 명시적으로 작성해주어야 읽는 사람이 쉽게 이해가 가능합니다.


또한 사실상 생성자가 필요없는 유틸성 클래스네요.
유틸리티 클래스의 경우 생성자를 private으로 변경하길 추천드립니다.
관련 글들을 찾아 학습해보세요 :)

xlffm3 commented on src/main/java/racingcar/validate/Validation.java:
```diff
@@ -0,0 +1,14 @@
+import java.util.List;
+
+public class Validation {
+	public void validateCarName(String inputCarName) {
+		List<String> carNameArray = Arrays.asList(inputCarName.split(","));
+		new CheckCarName(carNameArray);
+	}
+	public void validatePlayTime(int playTime) {
+		new CheckPlayTime(playTime);
+	}
```
Validator가 유효성을 체크해줄 것이라 생각했는데, CheckCarName, CheckPlayTime을 통해 각각 유효성을 검증하네요.

1. 연습하는 차원이니.. CheckCarName, CheckPlayTime 같은 클래스 명보다는 CarNameValidator 같이 명시적으로 작성해주는게 좋습니다.
2. CheckCarName.validate(carNameArray) 형태가 아니라 생성자를 통해 validate 하는 이유가 있을까요? '단일책임원칙'에 대해 알아보면 좋을 것 같아요. 생성자가 하는 역할, 책임이 뭘까요?
3. 객체끼리 메시지를 주고 받으며 협력하는게 객체지향이지만, 과연 'CheckCarName', 'CheckPlayTime' 이란 클래스가 필요할까요? 너무 과한 클래스 분리는 오히려 가독성을 해칠 수 있습니다. 이러한 tradeoff에 대해 고민해봐요.
4. 정말로 CheckCarName, CheckPlayTime이 필요하다면 내부에서 'new'를 통해 생성하기 보다는 Validation 클래스의 인스턴스 변수로 두고 생성자 주입을 받는 방법도 있을텐데요. 왜 'new'를 통해 생성했는지 이유가 궁금합니다~

jiyoung-Heo replied:
피드백을 해 주신 대로 고민을 많이 해 보며 수정하였습니다.
validator 클래스는 제거하였고 CarNameValidator, PlayTimeValidator 두개의 클래스는 싱글톤으로 만들었습니다. 
생성자는 인스턴스를 초기화하는 역할로 제가 만든 코드가 적합하지 않다는것을 인지하고 수정하였습니다.
new를 통해 생성하는것이 더욱 짧은 코드가 된다고 생각하여 그렇게 작성하였는데요, 이점은 싱글톤으로 바꾸게 되면서 수정하였습니다

xlffm3 commented on src/main/java/racingcar/validate/CheckPlayTime.java:
```diff
@@ -0,0 +1,18 @@
+import racingcar.constant.ConstNumber;
+
+public class CheckPlayTime {
+	public CheckPlayTime() {
+	}
+
+	public CheckPlayTime(int playTime) {
+		checkPlayTime(playTime);
+	}
+
```
1. 생성자를 통해 validate 하는 이유가 있을까요?
2. 사용하지 않는 기본 생성자를 만드신 이유가 있을까요?

jiyoung-Heo replied:
1. 생성자를 통해 한번에 실행시키기 위해 생성자를 통해 validate하였는데요,
이 점은 제가 생성자의 역할을 인지하지 못하고 잘못 작성하였다고 생각합니다. CarNAmeValidator와 마찬가지로 수정하였습니다.
2. 사용하지 않는 기본 생성자는 처음 코드를 작성할 때 사용할 수 있을지도 모른다는 생각에 만들어 두었는데 미처 삭제하지 못하였습니다.

xlffm3 commented on src/main/java/racingcar/validate/CheckPlayTime.java:
```diff
@@ -0,0 +1,18 @@
+		checkPlayTime(playTime);
+	}
+
+	public void checkPlayTime(int playTime) {
+		if (playTime < ConstNumber.PLAY_TIME_LEAST_COUNT) {
+			throw new IllegalArgumentException();
+		}
+	}
```
checkPlayTime 이란 메서드는 해당 클래스 내부에서만 사용할 뿐 외부에서는 사용하지 않습니다.
java는 접근 제어자를 제공하는데요. 외부에서 사용하지 않을 것을 public으로 열어두었을 때 단점이 뭘지 생각해보고 고쳐봅시다.

jiyoung-Heo replied:
네 외부에서 사용하지 않는 메서드는 모두 private으로 변경하였습니다.

public으로 열어두게 되면 보안 측면에서 자료보호가 불가능하다고 생각합니다.

xlffm3 commented on src/main/java/racingcar/validate/CheckCarName.java:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	public void checkCarNameLength(List<String> carNameArray){
+		for (String s : carNameArray) {
+			if (s.length() > ConstNumber.CAR_NAME_LENGTH ) {
+				throw new IllegalArgumentException();
+			}
+		}
```
지금은 연습 차원이니 코드의 depth를 1로 낮춰볼까요?
이럴 때 java stream api를 사용해보면 어떨까요?
그리고 지금 인자는 List인데 array라는 변수명이 어색합니다.

```java
carNames.stream()
.filter(carName -> carName.length() > 10)
.findAny()
            .ifPresent(t -> {
                throw new IllegalArgumentException();
            });
```

같은 방법이 있습니다.
allMatch, count 등의 다른 api도 있구요~

jiyoung-Heo replied:
stream api를 처음 사용해 봤는데 생소해서 어렵게 느껴졌습니다. 
혼자 코드를 구현하려다 실패해서 올려주신 코드를 참고하였습니다.!

다른 코드들도 depth를 1로 낮추려고 해봤지만 아직은 익숙하지 않아서 적용이 어렵더라구요
이 부분은 더 공부하고 적용시켜보면서 추후에 변경하도록 하겠습니다

xlffm3 commented on src/main/java/racingcar/validate/CheckCarName.java:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	public void checkCarNameDuplicate(List<String> carNameArray){
+		for (String s : carNameArray) {
+			if (Collections.frequency(carNameArray, s) > ConstNumber.CAR_NAME_DUPLICATE_COUNT) {
+				throw new IllegalArgumentException();
+			}
+		}
```
duplicate를 어떻게 확인할 수 있을까요?
자료 구조 중 Set을 활용해봅시다. (혹은 stream api를 통해서 해도 됩니다 ~~)

xlffm3 commented on src/main/java/racingcar/validate/CheckCarName.java:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	public void checkCarNameBlank(List<String> carNameArray){
+		for (String s : carNameArray) {
+			if (s.equals(" ")) {
+				throw new IllegalArgumentException();
+			}
+		}
```
이러한 확인 작업이 필요할까요?
예를 들어, 여러 공백이 들어온 자동차 이름은 걸러낼 수 없습니다.
'trim' 이라고 하는 java string api를 사용해봅시다.

jiyoung-Heo replied:
네 해당 부분 trim을 통해 걸러내고 stream api를 사용하는쪽으로 수정하였습니다.

xlffm3 commented on src/main/java/racingcar/validate/CheckCarName.java:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	public void checkCarNameEmpty(List<String> carNameArray){
+		for (String s : carNameArray) {
+			if (s.isEmpty()) {
+				throw new IllegalArgumentException();
+			}
+		}
```
carNameLength를 확인할 때 함께 확인이 가능할것같은데요.

jiyoung-Heo replied:
경우의 수를 최대한 많이 생각하고 많이 분리하자고 생각하다보니 겹치는 부분이발생했네요
해당부분 수정하였습니다!

xlffm3 commented on src/main/java/racingcar/validate/CheckCarName.java:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	public void checkCarNameLastComma(List<String> carNameArray){
+		if (carNameArray.get(carNameArray.size()-1).equals(",")) {
+			throw new IllegalArgumentException();
+		}
```
처음 자동차 이름을 입력받을 때 'pobi,abc,kevin,'과 같이 들어온 경우 걸러내려는 목적이군요.
차라리 처음에 정규표현식을 사용해보는건 어떨까요?

jiyoung-Heo replied:
처음에 정규표현식을 사용한다는것이 어떤 의미인지 이해하지 못했습니다 다시한번 설명해주실 수 있으실까요?
일단 해당 코드는 정규표현식으로 바꾸었습니다.

xlffm3 commented on src/main/java/racingcar/constant/ConstNumber.java:
```diff
@@ -0,0 +1,10 @@
+package racingcar.constant;
+
+public class ConstNumber {
+	public final static int CAR_NAME_LENGTH = 5;
+	public final static int CAR_NAME_DUPLICATE_COUNT = 1;
+	public final static int PLAY_TIME_LEAST_COUNT = 1;
+	public final static int RANDOM_NUMBER_MIN = 0;
+	public final static int RANDOM_NUMBER_MAX = 9;
+	public final static int MOVE_CONDITION_RANDOM_NUMBER = 4;
+}
```
특정 클래스가 혼자서 사용할 상수를 외부 클래스에 정의해두고 public으로 열어두고 있습니다.
하나의 클래스만 사용할 상수라면 해당 클래스 내부에 정의해두는 것을 권장합니다.

만약 나중에 기능이 추가되서 CAR_NAME_LENGTH 상수를 의존하는 클래스가 많아지게 되면 어떻게 될까요?
일부 클래스들에 대한 정책이 변경되어 CAR_NAME_LENGTH를 10으로 변경하는 순간, 5 값에 의존하던 클래스들은 예기치 못한 변경으로 인해 영향을 받게 되고 리팩터링이 어려워집니다.

jiyoung-Heo replied:
마지막에 설명하신 부분은 생각해보지 못한 부분이었습니다. 말씀대로 리팩터링이 어려워지네요 하나로 모아두면 관리하기 편하다고 생각하였는데 그게 아니었네요,, 모두 하나의 클래스에서만 사용하는 상수이기에 클래스 내부에 privite으로 정의하도록 수정하였습니다

xlffm3 commented on src/main/java/racingcar/InputUser.java:
```diff
@@ -0,0 +1,43 @@
+import racingcar.validate.Validation;
+
+public class InputUser {
+	Validation validation = new Validation();
+
```
1. default 접근 제어자를 사용한 이유가 있을까요?
2. 객체를 외부에서 생성자로 주입받지 않고 내부에서 생성하는 이유가 있을까요?

jiyoung-Heo replied:
1. 접근제어자에 대해 제대로 인지하고 있지 못하여서 이렇게 사용하였습니다.. 올바른 접근제어자를 넣어주도록 하겠습니다
2. 해당 객체가 다양한 곳에 쓰이지 않고 하나의 클래스에만 쓰이기 때문에 굳이 외부에서 생성자로 주입받게 하지 않았습니다.!

xlffm3 commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,65 @@
+import racingcar.constant.ConstMessage;
+
+public class RacingGame {
+	List<Car> carList = new LinkedList<>();
```
LinkedList, default 생성자를 사용한 이유가 있을까요?

jiyoung-Heo replied:
LinkedList의 경우 처음 생성할 때 삭제나 삽입이 있을것으로 예상하여 생성하였습니다.
그러나 현재 기능상 ArrayList가 더 적절한것으로 판단하여 ArrayList로 수정하였습니다

xlffm3 commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,65 @@
+		}
+	}
+
+	public List<String> getWinner(List<Car> carList) {
+		List<String> winner = new ArrayList<>();
+		for (Car car : carList) {
+			if (car.getPosition() == getMaxPosition()) {
+				winner.add(car.getName());
+			}
+		}
+		return winner;
+	}
```
for, if depth를 1로 줄여봅시다. 이 부분도 stream으로 처리할 수 있을 것 같네요.

Car에서 position을 getter로 가져와 비교하고 잇는데, 이러한 부분을 Car 객체에게 메시지를 보내도록 수정해보세요.

또한 Car List를 Cars라고 하는 클래스로 랩핑하면 승자를 구하는 로직을 해당 객체에게 위임할 수 있겠네요.

jiyoung-Heo replied:
stream api를 통해 depth를 1로 수정하였습니다.

Car 객체에서 isMaxPosition 메소드를 생성하여 boolean을 리턴하도록 만들었습니다. 

Cars리스트를 생성하여 Carlist, maxposition 메소드를 만들어서 리스트와 승자를 구하는 로직을 담당하도록 하였습니다. 

Cars, Car, RacingGame의 메서드들이 알맞게 분리되어있지 않다고 생각됩니다. 이부분은 더 고민해보겠습니다.

xlffm3 commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,65 @@
+		return winner;
+	}
+
+	public int getMaxPosition() {
+		int maxPosition = 0;
+
+		for (Car car : carList) {
+			if (maxPosition < car.getPosition()) {
+				maxPosition = car.getPosition();
+			}
+		}
+
+		return maxPosition;
```
클래스 내부에서만 사용할 메서드가 public으로 열려있습니다.

xlffm3 commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,65 @@
+	public void printWinner(List<String> winner) {
+		System.out.print(ConstMessage.PRINT_MESSAGE_FINAL_WINNER);
+
+		for (int i = 0; i < winner.size(); i++) {
+			System.out.print(winner.get(i));
+			if (winner.size() != 1 && i != winner.size() - 1) {
+				System.out.print(", ");
+			}
+		}
```
String method나 Stream API에 쉼표를 삽입해주는 기능이 있습니다.
이러한 부분을 학습하고 적용해보시면 좋을 것 같아요.

jiyoung-Heo replied:
String.join()메서드를 통해 쉼표를 삽입해 주었습니다.

제가 너무 주먹구구식으로 코드를 짰다는 생각이 들었습니다. 다양한 메서드들을 사용하고 익혀야겠습니다

xlffm3 reviewed (changes requested):
안녕하세요 @jiyoung-Heo 님.
예외 사항에 대해 꼼꼼하게 코드를 작성해주셨네요. 👍👍
몇 가지 피드백을 남겨드렸어요. 

코드 전반적으로 method depth가 2 이상입니다. 어렵더라도 1로 줄여보는 연습을 해봅시다.
해당 피드백들을 반영하고 다시 리뷰 요청해주세요.


---

(선택)
만약 시간이 남는다면 MVC 패턴을 학습해보고 적용해봅시다.
현재 콘솔에 print하는 메서드가 Car, RacingGame 등등 비즈니스 로직을 가진 핵심 도메인에 산재되어있습니다. 이렇게 산재되어있으면 어떤 단점이 있을까요?

또한 객체의 '책임'에 대해 고민해보시면 좋을것같아요. 'Car' 객체의 책임은 움직이는 것이지, 정보를 출력하는게 'Car' 객체의 책임일까요?

xlffm3 commented on src/main/java/racingcar/Car.java:
```diff
@@ -1,12 +1,44 @@
+	private final String name;
+	private int position = 0;
+
+	public Car(String name) {
+		this.name = name;
+	}
```
Car 객체를 생성할 때에는 이름에 대한 Length 검사가 없는 이유가 있을까요?

jiyoung-Heo replied:
InputUser에서 값을 입력받을 때 검사하였기 때문에 따로 검사하지 않았습니다

xlffm3 replied:
View에서 유효성 검사를 진행한다고 Model에서 유효성 검사를 진행하지 않아도 될까요? 자동차의 이름이 null, empty, 적정한 길이가 아닌지 등을 확인하는 것은 Car 객체의 고유한 책임이 아닐까요? 만약 유효성 검사를 두 곳에서 진행하기 어렵다면 적어도 모델 쪽에서 진행하는 것이 낫다고 생각합니다 :)

---

https://jojoldu.tistory.com/157

웹 개발 글이긴 한데요. 프론트엔드에서 입력값에 대한 유효성 검사를 진행하더라도 백엔드 차원에서도 유효성 검사를 진행해주어야합니다. 물론 개발 프로세스, 상황에 따라 항상 양쪽 모두 검사하기 어려운 부분도 있겠습니다만...

jiyoung-Heo committed:
docs(README.md): 피드백 후 수정목록 작성

jiyoung-Heo committed:
style(Validator): 클래스 명 변경

Validation에서 Validator로 변경

jiyoung-Heo committed:
style(carNameList): 변수명 변경

carNameArray에서 carNameList로 변경

jiyoung-Heo committed:
refactor(checkCarNameLength): checkCarNameEmpty 메소드와 합치기

checkCarNameEmpty 에서 체크하던 부분을 checkCarNameLength 에서 함께 확인가능

jiyoung-Heo committed:
refactor(checkCarNameLastComma): 정규표형식 사용하여 문자열 끝의 , 거르기

jiyoung-Heo committed:
refactor: 상수 클래스 분리하지 않고 해당 클래스 내부에 정의

jiyoung-Heo committed:
refactor(checkCarNameBlank): trim을 통해 여러개의 공백을 걸러주기

.equals(" ")사용하면 두개의 공백이 들어온 경우 걸러낼 수 없음

jiyoung-Heo committed:
refactor: 클래스 내부에서만 사용하는 메서드 private으로 변경

jiyoung-Heo committed:
refactor(printWinner): for문으로 쉼표 추가하지 않고 String.join()메소드를 통해 쉼표 삽입하기

jiyoung-Heo committed:
refactor(checkCarNameDuplicate): stream api를 이용하여 duplicate 확인

jiyoung-Heo committed:
refactor(validation): 접근제어자 private으로 변경

jiyoung-Heo committed:
refactor(CheckCarName,CheckPlayTime): 기본생성자 삭제

jiyoung-Heo committed:
refactor: Car isMaxPosition 생성

jiyoung-Heo committed:
refactor(getWinner): depth 1로 줄이기 stream

jiyoung-Heo committed:
refactor: 구조 수정

Cars클래스 생성해서 RacingGame의 메소드 분리
Validator 클래스 삭제 후 PlayTimeValidator, CarNameValidator 두 클래스로 예외처리

jiyoung-Heo committed:
refactor(PlayTimeValidator, CarNameValidator): 싱글톤으로 바꾸기

jiyoung-Heo committed:
refactor: 접근제한자 정리

jiyoung-Heo committed:
style: 사용하지 않는 import 제거

jiyoung-Heo committed:
refactor(carList): LinkedList에서 ArrayList로 변경

처음 생성할 때 삭제나 삽입이 있을것으로 예상하여 생성하게 되었는데 현재 기능상 ArrayList가 더 적절한것으로 판단하여 수정

jiyoung-Heo committed:
style: 코드 간격 수정

jiyoung-Heo commented:
리뷰 남겨주신 부분에 대해 수정완료하였습니다. 

depth 줄이는 부분은 최대한 해 보았는데 아직 줄이지 못한 곳이 많았습니다. 이후에도 계속 stream api를 접해보며 많은 코드를 작성해보겠습니다!
mvc패턴같은 경우 시간이 부족하여 적용하지 못하였습니다. 이 부분은 추후에 적용해 보도록 하겠습니다

일주일이라는 시간이 이렇게 짧을 줄 몰랐는데 피드백 주신 부분 학습하다 보니 훌쩍 지나가버렸습니다. 스스로 부족함을 너무너무 느끼게 되었고 더욱 노력하고싶은 마음이 샘솟았습니다 ㅎㅎ
다시한번 리뷰 요청드립니다
감사합니다!

xlffm3 commented on src/main/java/racingcar/util/CarNameValidator.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.List;
+import java.util.regex.Pattern;
+
+public class CarNameValidator {
+	private final static int CAR_NAME_LENGTH = 5;
+
+	private CarNameValidator() {
+	}
+
+	private static CarNameValidator instance = new CarNameValidator();
```
자바 컨벤션대로, 상수 부분은 클래스 최상단쪽에 순서를 지켜서 정의해주세요.
그리고 instance 또한 final로 정의하고 상수화시켜주세요~

xlffm3 commented on src/main/java/racingcar/util/PlayTimeValidator.java:
```diff
@@ -0,0 +1,20 @@
+package racingcar.util;
+
+public class PlayTimeValidator {
+	private final static int PLAY_TIME_LEAST_COUNT = 1;
+
+	private PlayTimeValidator() {
+	}
+
+	private static PlayTimeValidator instance = new PlayTimeValidator();
```
이 부분도 컨벤션 순서가 안 맞네요~

xlffm3 commented on src/main/java/racingcar/util/CarNameValidator.java:
```diff
@@ -0,0 +1,57 @@
+	}
+
+	private void checkCarNameLastComma(List<String> carNameList){
+		if(Pattern.matches("/,$/", carNameList.toString())){
+			throw new IllegalArgumentException();
+		}
```
정규 표현식 잘 사용해주셨네요. 다만 Pattern 객체는 매번 생성하는 것이 비용이 비싸답니다.

다양한 regex api들은 내부적으로 Pattern을 생성해서 씁니다. 미리 정적 상수로 컴파일해두는게 좋습니다!

xlffm3 commented on src/main/java/racingcar/Cars.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.stream.Collectors;
+
+public class Cars {
+	private List<Car> carList = new ArrayList<>();
```
1. 인스턴스 변수와 클래스 선언 부분 사이는 1칸 띄워주세요~
2. final로 선언할 수 있다면 final을 해주세요~
3. 변수명에 list와 같은 구체적인 자료형을 선언해주기 보다는 다른 메서드 네이밍을 사용해봅시다. 가령, 리스트에서 다른 자료형으로 변경되면 변수명까지 함께 변경되어야하기 때문입니다.
4. 아울러 해당 부분을 내부에서 new ArrayList를 생성하는 것 보다 Cars 인스턴스 생성시 생성자로 주입받는 것을 추천드려요. 의존성 주입이 왜 중요한지에 대해 공부해보면 좋겠습니다 :)

xlffm3 reviewed (approved):
@jiyoung-Heo
안녕하세요~~ 피드백이 조금 늦었습니다. 코드 리뷰 남겨드린 내용이 상당히 많았는데 꼼꼼하게 공부하시고 반영해주시려는 모습 너무 좋았습니다!! 특히 코드들도 많이 개선됬네요 ㅎㅎ 이만하면 이번 미션은 충분하다고 생각되어 approve 드립니다.

(그런데 conflict가 있어서 머지가 안 되네요~ 이점 확인 부탁드려요!)

또 궁금하신부분 있으면 언제든 멘션주세요.

jiyoung-Heo commented:
피드백 감사합니다 :) 피드백 반영하여 더 공부해나가겠습니다!!

메인브랜치에 다른유저가 푸시한 이력이 있어서 conflict 나는것 같습니다. 
다른 브랜치도 있는데요 ,,, 어떻게 해결해야할지 모르겠습니다.

jiyoung-Heo commented:
conflict 나는 것에 대해 구글링 해 보았는데 해결방법을 찾지 못했습니다...
한번 확인해 주실 수 있나요..?
